File: internal/core/lifecycle_test.go
// Fiscal convention: monetary values in the Drenyra ecosystem are int64 cents;
// no float is ever used for money. This module tests the v2 lifecycle machine
// (approval gate, void, supersession) against the SQLite store, the same way the
// shared API composes it: core decides, the store persists.

package core_test

import (
	"path/filepath"
	"strings"
	"testing"

	"github.com/arkelythex/drenyra-engram/internal/core"
	"github.com/arkelythex/drenyra-engram/internal/store"
)

const testTimestamp = "2026-01-15T12:00:00.000Z"

var testScope = core.Scope{
	Kind:           core.ScopeKindCompany,
	OrganizationID: "org-001",
	CompanyID:      "acme",
	RUC:            "20100039201",
	Period:         "202401",
}

var testContent = core.Content{
	What:    "IGV base rate is 18 percent",
	Why:     "standard rate for goods",
	Where:   "Peru",
	Learned: "applies to all invoices",
}

var testSource = core.Source{
	System:    "go-test",
	ActorID:   "test-agent",
	ActorKind: core.ActorKindAgent,
}

var humanSource = core.Source{
	System:    "go-test",
	ActorID:   "maria.torres",
	ActorKind: core.ActorKindHuman,
}

var systemSource = core.Source{
	System:    "drenyra-core",
	ActorID:   "scheduler",
	ActorKind: core.ActorKindSystem,
}

func newTestStore(t *testing.T) *store.SQLiteStore {
	t.Helper()
	path := filepath.Join(t.TempDir(), "engram.db")
	s, err := store.Open(path)
	if err != nil {
		t.Fatalf("open test store: %v", err)
	}
	t.Cleanup(func() { _ = s.Close() })
	return s
}

func saveInput(what string) core.SaveInput {
	return core.SaveInput{
		TopicKey:     "tax.igv.rate",
		Title:        "IGV base rate",
		Kind:         core.KindRule,
		Scope:        testScope,
		Content:      core.Content{What: what, Why: "standard rate for goods", Where: "Peru", Learned: "applies to all invoices"},
		FiscalEffect: core.FiscalEffectNone,
		EffectiveAt:  testTimestamp,
		Source:       testSource,
	}
}

// TestSaveGateStatus is the v2 save gate: fiscalEffect none → active (no gate),
// any non-none effect → pending_review (mandatory human approval).
func TestSaveGateStatus(t *testing.T) {
	s := newTestStore(t)

	informative, err := s.Save(saveInput("informative version"))
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	if informative.Memory.Status != core.StatusActive {
		t.Fatalf("status = %s, want active for fiscalEffect none", informative.Memory.Status)
	}

	gated := saveInput("gated version")
	gated.TopicKey = "tax.igv.adjustment"
	gated.Kind = core.KindDecision
	gated.FiscalEffect = core.FiscalEffectAdjustment
	gated.EffectiveAt = "2026-01-31"
	review, err := s.Save(gated)
	if err != nil {
		t.Fatalf("save gated: %v", err)
	}
	if review.Memory.Status != core.StatusPendingReview {
		t.Fatalf("status = %s, want pending_review for fiscalEffect adjustment", review.Memory.Status)
	}
	if !core.IsGated(gated.FiscalEffect) {
		t.Fatal("IsGated(adjustment) = false, want true")
	}
}

// TestApproveHumanFlow: a pending_review memory reaches approved through a HUMAN
// actor, and the audit trail records actor + actorKind.
func TestApproveHumanFlow(t *testing.T) {
	s := newTestStore(t)
	saved, err := s.Save(gatedInput())
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	id := saved.Memory.Identity.ID

	fromStore, ok := s.FindByID(id)
	if !ok {
		t.Fatal("memory not found")
	}
	if err := core.Approve(&fromStore, humanSource); err != nil {
		t.Fatalf("approve: %v", err)
	}
	approved, err := s.ApplyStatusTransition(id, core.StatusApproved, core.TransitionMeta{
		Actor:     humanSource.ActorID,
		ActorKind: humanSource.ActorKind,
		Timestamp: testTimestamp,
	})
	if err != nil {
		t.Fatalf("applyStatusTransition: %v", err)
	}
	if approved.Status != core.StatusApproved {
		t.Fatalf("status = %s, want approved", approved.Status)
	}
	fromStore, ok = s.FindByID(id)
	if !ok || fromStore.Status != core.StatusApproved {
		t.Fatalf("stored status = %v, want approved", fromStore.Status)
	}

	log, err := s.TransitionLog()
	if err != nil {
		t.Fatalf("transition log: %v", err)
	}
	found := false
	for _, entry := range log {
		if entry.MemoryID == id &&
			entry.From == core.StatusPendingReview &&
			entry.To == core.StatusApproved &&
			entry.Actor == "maria.torres" &&
			entry.ActorKind == core.ActorKindHuman {
			found = true
		}
	}
	if !found {
		t.Fatalf("audit-trail entry missing: %+v", log)
	}
}

// TestApproveByAgentFailsClosed: the gate is non-negotiable — an agent CANNOT
// approve, and the stored memory stays pending_review.
func TestApproveByAgentFailsClosed(t *testing.T) {
	s := newTestStore(t)
	saved, err := s.Save(gatedInput())
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	id := saved.Memory.Identity.ID

	fromStore, _ := s.FindByID(id)
	approveErr := core.Approve(&fromStore, testSource)
	if approveErr == nil {
		t.Fatal("agent approval must fail closed")
	}
	if !strings.Contains(approveErr.Error(), core.ErrGateRequiresHuman.Error()) {
		t.Fatalf("expected GATE_REQUIRES_HUMAN, got %v", approveErr)
	}
	fromStore, _ = s.FindByID(id)
	if fromStore.Status != core.StatusPendingReview {
		t.Fatalf("memory changed after failed gate: %s", fromStore.Status)
	}
}

// TestRejectHumanFlow: a human can reject a pending_review memory (terminal).
func TestRejectHumanFlow(t *testing.T) {
	s := newTestStore(t)
	saved, err := s.Save(gatedInput())
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	id := saved.Memory.Identity.ID

	fromStore, _ := s.FindByID(id)
	if err := core.Reject(&fromStore, humanSource); err != nil {
		t.Fatalf("reject: %v", err)
	}
	rejected, err := s.ApplyStatusTransition(id, core.StatusRejected, core.TransitionMeta{
		Actor:     humanSource.ActorID,
		ActorKind: humanSource.ActorKind,
		Timestamp: testTimestamp,
	})
	if err != nil {
		t.Fatalf("applyStatusTransition: %v", err)
	}
	if rejected.Status != core.StatusRejected {
		t.Fatalf("status = %s, want rejected", rejected.Status)
	}

	// Terminal: rejecting again is an illegal transition and leaves the store
	// unchanged.
	fromStore, _ = s.FindByID(id)
	if err := core.Reject(&fromStore, humanSource); err == nil || !strings.Contains(err.Error(), core.ErrInvalidTransition.Error()) {
		t.Fatalf("double reject: expected INVALID_TRANSITION, got %v", err)
	}
	if _, err := s.ApplyStatusTransition(id, core.StatusRejected, core.TransitionMeta{Actor: "x", Timestamp: testTimestamp}); err != nil {
		t.Fatalf("persisting the illegal transition must still be a store-level error-free op (legality is the machine's job), got: %v", err)
	}
}

// TestVoidFlow: void admits human/system actors, never agents; result is
// terminal.
func TestVoidFlow(t *testing.T) {
	s := newTestStore(t)

	t.Run("agent cannot void", func(t *testing.T) {
		saved, err := s.Save(saveInput("to-void"))
		if err != nil {
			t.Fatalf("save: %v", err)
		}
		id := saved.Memory.Identity.ID
		m, _ := s.FindByID(id)
		if err := core.Void(&m, testSource); err == nil || !strings.Contains(err.Error(), "GATE_AGENT_CANNOT_VOID") {
			t.Fatalf("expected GATE_AGENT_CANNOT_VOID, got %v", err)
		}
		if stored, _ := s.FindByID(id); stored.Status != core.StatusActive {
			t.Fatalf("memory mutated by failed void: %s", stored.Status)
		}
	})

	t.Run("system void succeeds", func(t *testing.T) {
		saved, err := s.Save(saveInput("to-void"))
		if err != nil {
			t.Fatalf("save: %v", err)
		}
		id := saved.Memory.Identity.ID
		m, _ := s.FindByID(id)
		if err := core.Void(&m, systemSource); err != nil {
			t.Fatalf("system void: %v", err)
		}
		updated, err := s.ApplyStatusTransition(id, core.StatusVoided, core.TransitionMeta{
			Actor:     systemSource.ActorID,
			ActorKind: systemSource.ActorKind,
			Timestamp: testTimestamp,
		})
		if err != nil {
			t.Fatalf("applyStatusTransition: %v", err)
		}
		if updated.Status != core.StatusVoided {
			t.Fatalf("status = %s, want voided", updated.Status)
		}
	})
}

// TestSupersedePrevViaSave: a new save of the same (topicKey, scope) chain marks
// the previous current revision superseded with the successor id — history is
// immutable, readers route onward.
func TestSupersedePrevViaSave(t *testing.T) {
	s := newTestStore(t)
	first, err := s.Save(saveInput("first version"))
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	second, err := s.Save(saveInput("second version"))
	if err != nil {
		t.Fatalf("save: %v", err)
	}

	old, ok := s.FindByID(first.Memory.Identity.ID)
	if !ok {
		t.Fatal("revision 1 lost")
	}
	if old.Status != core.StatusSuperseded {
		t.Fatalf("status = %s, want superseded", old.Status)
	}
	if old.SupersedesID != second.Memory.Identity.ID {
		t.Fatalf("SupersedesID = %q, want %q", old.SupersedesID, second.Memory.Identity.ID)
	}
	if old.Content.What != "first version" || old.Revision != 1 {
		t.Fatalf("superseded revision mutated: %+v", old)
	}

	latest, ok := s.FindByTopicKey("tax.igv.rate", testScope)
	if !ok || latest.Revision != 2 {
		t.Fatalf("latest = %+v, want revision 2", latest)
	}
}

// TestIllegalTransitionLeavesStoredMemoryUnchanged: a transition the machine
// rejects never mutates the stored row.
func TestIllegalTransitionLeavesStoredMemoryUnchanged(t *testing.T) {
	s := newTestStore(t)
	saved, err := s.Save(saveInput("first version"))
	if err != nil {
		t.Fatalf("save: %v", err)
	}
	id := saved.Memory.Identity.ID

	m, _ := s.FindByID(id)
	if err := core.Approve(&m, humanSource); err == nil || !strings.Contains(err.Error(), core.ErrInvalidTransition.Error()) {
		t.Fatalf("approve from active: expected INVALID_TRANSITION, got %v", err)
	}
	stored, _ := s.FindByID(id)
	if stored.Status != core.StatusActive {
		t.Fatalf("stored status = %s, want unchanged active", stored.Status)
	}
}

func gatedInput() core.SaveInput {
	return core.SaveInput{
		TopicKey:     "tax.igv.adjustment",
		Title:        "Ajuste de IGV",
		Kind:         core.KindDecision,
		Scope:        testScope,
		Content:      testContent,
		FiscalEffect: core.FiscalEffectJournalEntry,
		EffectiveAt:  "2026-01-31",
		Source:       testSource,
	}
}

File: lifecycle/__tests__/lifecycle.test.ts
/**
 * Fiscal convention: monetary values in the Drenyra ecosystem are BigInt cents;
 * no float is ever used for money; sequence/revision counters are JSON integers,
 * never floats.
 *
 * Lifecycle v2 (contracts/lifecycle.md):
 * - The human approval gate: a memory with fiscalEffect != none lands
 *   `pending_review`; only a HUMAN actor can approve/reject it
 *   (GATE_REQUIRES_HUMAN otherwise). Informative memory (fiscalEffect none)
 *   saves directly `active`.
 * - approve/reject/void/supersede legality fail closed (INVALID_TRANSITION).
 * - Vigencia: an expired memory (expiresAt < now) surfaces as stale in search
 *   results, never as current fact.
 */

import { describe, expect, it } from "vitest";

import type { MemoryContent, MemoryScope } from "../../core/types.js";
import { InMemoryMemoryStore } from "../../store/memory-store.js";
import { scopeFirstSearch } from "../../search/scope-first.js";
import {
	applyGateTransition,
	approve,
	assertHumanApproval,
	canApprove,
	canVoid,
	initialStatus,
	isGated,
	reject,
	voidMemory,
} from "../transitions.js";

const SCOPE: MemoryScope = {
	kind: "company",
	organizationId: "org-001",
	companyId: "acme",
	ruc: "20100039201",
	period: "202401",
};
const T = "2026-01-15T12:00:00.000Z";

const CONTENT: MemoryContent = {
	what: "IGV base rate is 18 percent",
	why: "standard rate for goods",
	where: "Peru",
	learned: "applies to all invoices",
};

const agentMeta = {
	actor: "test-agent",
	actorKind: "agent" as const,
	timestamp: T,
};
const humanMeta = {
	actor: "maria.torres",
	actorKind: "human" as const,
	timestamp: T,
};

const informative = {
	topicKey: "tax.igv.rate",
	title: "IGV base rate",
	kind: "rule" as const,
	scope: SCOPE,
	content: CONTENT,
	fiscalEffect: "none" as const,
	effectiveAt: "2024-01-01T00:00:00.000Z",
	source: {
		system: "vitest",
		actorId: "test-agent",
		actorKind: "agent" as const,
	},
};

const gated = {
	...informative,
	topicKey: "adjust/aj-001",
	title: "Ajuste AJ-001",
	kind: "decision" as const,
	fiscalEffect: "adjustment" as const,
	effectiveAt: "2024-01-31T00:00:00.000Z",
};

describe("the approval gate", () => {
	it("initialStatus: none → active; any fiscal effect → pending_review", () => {
		expect(initialStatus("none")).toBe("active");
		expect(initialStatus("adjustment")).toBe("pending_review");
		expect(initialStatus("journal_entry")).toBe("pending_review");
		expect(initialStatus("sunat_filing")).toBe("pending_review");
		expect(isGated("none")).toBe(false);
		expect(isGated("closing")).toBe(true);
	});

	it("a memory with fiscal effect is saved pending_review", async () => {
		const store = new InMemoryMemoryStore();
		const saved = await store.save(gated);
		expect(saved.memory.status).toBe("pending_review");
	});

	it("only pending_review memories can be approved", () => {
		expect(canApprove("pending_review")).toBe(true);
		expect(canApprove("active")).toBe(false);
		expect(canApprove("approved")).toBe(false);
		expect(canApprove("superseded")).toBe(false);
	});

	it("a machine cannot approve — GATE_REQUIRES_HUMAN", () => {
		const memory = {
			...gated,
			identity: { id: "m1", topicKey: gated.topicKey },
			status: "pending_review" as const,
		};
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		expect(() => approve(memory as any, agentMeta)).toThrow(
			/GATE_REQUIRES_HUMAN/,
		);
		expect(() => assertHumanApproval("agent")).toThrow(/GATE_REQUIRES_HUMAN/);
		expect(() => assertHumanApproval("human")).not.toThrow();
	});

	it("a human approves a pending_review memory through the store", async () => {
		const store = new InMemoryMemoryStore();
		const saved = await store.save(gated);
		const approved = applyGateTransition(
			store,
			saved.memory.identity.id,
			"approved",
			humanMeta,
		);
		expect(approved.status).toBe("approved");
		const log = store.transitionLog();
		expect(log).toHaveLength(1);
		expect(log[0].memoryId).toBe(saved.memory.identity.id);
		expect(log[0].from).toBe("pending_review");
		expect(log[0].to).toBe("approved");
		expect(log[0].actorKind).toBe("human");
	});

	it("approving an active (informative) memory fails closed with INVALID_TRANSITION", async () => {
		const store = new InMemoryMemoryStore();
		const saved = await store.save(informative);
		expect(saved.memory.status).toBe("active");
		expect(() =>
			applyGateTransition(
				store,
				saved.memory.identity.id,
				"approved",
				humanMeta,
			),
		).toThrow(/INVALID_TRANSITION/);
	});

	it("a machine cannot reject — GATE_REQUIRES_HUMAN; a human can", async () => {
		const store = new InMemoryMemoryStore();
		const saved = await store.save(gated);
		const memory = { ...saved.memory };
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		expect(() => reject(memory as any, agentMeta)).toThrow(
			/GATE_REQUIRES_HUMAN/,
		);
		const rejected = applyGateTransition(
			store,
			saved.memory.identity.id,
			"rejected",
			humanMeta,
		);
		expect(rejected.status).toBe("rejected");
		// Terminal: a rejected memory cannot be approved later.
		expect(() =>
			applyGateTransition(
				store,
				saved.memory.identity.id,
				"approved",
				humanMeta,
			),
		).toThrow(/INVALID_TRANSITION/);
	});

	it("void admits human or system, never an agent", async () => {
		expect(canVoid("active")).toBe(true);
		expect(canVoid("approved")).toBe(true);
		expect(canVoid("superseded")).toBe(false);
		expect(canVoid("voided")).toBe(false);
		const memory = {
			...gated,
			identity: { id: "m1", topicKey: gated.topicKey },
			status: "active" as const,
		};
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		expect(() => voidMemory(memory as any, agentMeta)).toThrow(
			/GATE_AGENT_CANNOT_VOID/,
		);
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		expect(() => voidMemory(memory as any, humanMeta)).not.toThrow();
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		expect(() =>
			voidMemory(memory as any, {
				actor: "system",
				actorKind: "system" as const,
				timestamp: T,
			}),
		).not.toThrow();
	});
});

describe("supersession (immutable history)", () => {
	it("a second save of the same chain supersedes the first atomically", async () => {
		const store = new InMemoryMemoryStore();
		const first = await store.save(informative);
		const second = await store.save({
			...informative,
			content: { ...CONTENT, what: "v2" },
		});

		const storedFirst = store.findById(first.memory.identity.id);
		expect(storedFirst?.status).toBe("superseded");
		expect(storedFirst?.supersedesId).toBe(second.memory.identity.id);
		expect(store.successorOf(first.memory.identity.id)?.identity.id).toBe(
			second.memory.identity.id,
		);
		expect(store.listByStatus("active")).toHaveLength(1);
		expect(store.listByStatus("superseded")).toHaveLength(1);
	});
});

describe("vigencia (stale marking)", () => {
	it("an expired memory surfaces as stale in search, never as current fact", async () => {
		const store = new InMemoryMemoryStore();
		await store.save({
			...informative,
			topicKey: "tax.igv.expired",
			validity: {
				effectiveAt: "2020-01-01T00:00:00.000Z",
				expiresAt: "2020-06-01T00:00:00.000Z",
			},
		});

		const results = scopeFirstSearch(store, {
			query: "IGV base rate",
			scope: SCOPE,
		});
		expect(results).toHaveLength(1);
		expect(results[0].stale).toBe(true);
	});
});
